### ISSUE TITLE
Order matters when retrieving services from the ContainerBuilder 

### ISSUE THREAD

**Issue by RonRademaker** [symfony/symfony#19362]
In some cases the order in which services are retrieved (and thus constructed) from the ContainerBuilder matter. I've included an example where the order in which services are retrieved make the difference between getting the service as defined and triggering a `Symfony\Component\DependencyInjection\Exception\ServiceCircularReferenceException`.

Reproduction code, this works as 'expected', resulting in both a `Foo` and `Bar` object in their respective variables.

``` php
<?php

use Symfony\Component\DependencyInjection\ContainerBuilder;
use Symfony\Component\DependencyInjection\Definition;
use Symfony\Component\DependencyInjection\Reference;

$containerBuilder = new ContainerBuilder();

$definition = new Definition(Foobar::class, [new Reference('foo')]);
$containerBuilder->setDefinition('foobar', $definition);

$definition = new Definition(Foo::class, [new Reference('bar')]);
$containerBuilder->setDefinition('foo', $definition);

$definition = new Definition(Bar::class);
$definition->addMethodCall('addFoobar', [new Reference('foobar')]);
$containerBuilder->setDefinition('bar', $definition);

$bar = $containerBuilder->get('bar');
$foo = $containerBuilder->get('foo');
```

This code triggers a `ServiceCircularReferenceException`.

``` php
<?php

use Symfony\Component\DependencyInjection\ContainerBuilder;
use Symfony\Component\DependencyInjection\Definition;
use Symfony\Component\DependencyInjection\Reference;

$containerBuilder = new ContainerBuilder();

$definition = new Definition(Foobar::class, [new Reference('foo')]);
$containerBuilder->setDefinition('foobar', $definition);

$definition = new Definition(Foo::class, [new Reference('bar')]);
$containerBuilder->setDefinition('foo', $definition);

$definition = new Definition(Bar::class);
$definition->addMethodCall('addFoobar', [new Reference('foobar')]);
$containerBuilder->setDefinition('bar', $definition);

$foo = $containerBuilder->get('foo');
```

**Comment by sstok** [comment]
When using the ContainerBuilder you must first `compile()` it before it's safe to rely on any service to work properly. Which is also why it's no longer supported to load a Service from an uncompiled container: https://github.com/symfony/symfony/pull/18728

**Comment by RonRademaker** [comment]
Adding ```$containerBuilder->compile();``` after ```$containerBuilder->setDefinition('bar', $definition);``` makes no difference.

**Comment by xabbuh** [comment]
Reopening here as #21436 was reverted until we have a solution that works on `master` too (we are no able to do that because of #21639 and I will resubmit #21436 right away).

**Comment by stof** [comment]
@xabbuh the order matters here, because the cycle involves constructor arguments. Cases where only methods calls are involved (or constructor arguments only in definitions inlined inside a method call) are totally fine (and this is precisely what ``services9`` is checking in the testsuite).
So if you work on hardening the detection, please ensure that this use case keeps working (meaning, don't change existing tests involving ``services9``)

**Comment by nicolas-grekas** [comment]
Fixed in #24822
